runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java

```diff
@@ -50,56 +51,61 @@
 public class StateRequestHandlers {
 
   /**
-   * A handler for multimap side inputs.
+   * A handler for side inputs.
    *
    * <p>Note that this handler is expected to be thread safe as it will be invoked concurrently.
    */
   @ThreadSafe
-  public interface MultimapSideInputHandler<K, V, W extends BoundedWindow> {
+  public interface SideInputHandler<V, W extends BoundedWindow> {
     /**
      * Returns an {@link Iterable} of values representing the side input for the given key and
      * window.
      *
+     * <p>The key is interpreted according to the access pattern of side input.
+     *
      * <p>TODO: Add support for side input chunking and caching if a {@link Reiterable} is returned.
      */
-    Iterable<V> get(K key, W window);
+    Iterable<V> get(byte[] key, W window);
+
+    /** Returns the {@link Coder} to use for the elements of the resulting values iterable. */
+    Coder<V> resultCoder();
   }
 
   /**
-   * A factory which constructs {@link MultimapSideInputHandler}s.
+   * A factory which constructs {@link SideInputHandler}s.
    *
    * <p>Note that this factory should be thread safe because it will be invoked concurrently.
    */
   @ThreadSafe
-  public interface MultimapSideInputHandlerFactory {
+  public interface SideInputHandlerFactory {
 
     /**
-     * Returns a {@link MultimapSideInputHandler} for the given {@code pTransformId} and {@code
-     * sideInputId}. The supplied {@code keyCoder}, {@code valueCoder}, and {@code windowCoder}
-     * should be used to encode/decode their respective values.
+     * Returns a {@link SideInputHandler} for the given {@code pTransformId}, {@code sideInputId},
+     * and {@code accessPattern}. The supplied {@code elementCoder} and {@code windowCoder} should
+     * be used to encode/decode their respective values.
      */
-    <K, V, W extends BoundedWindow> MultimapSideInputHandler<K, V, W> forSideInput(
+    <T, V, W extends BoundedWindow> SideInputHandler<V, W> forSideInput(
```

Comment:
Instead of trying to force two different access patterns into a single interface that sort of works for both by special casing the key, wouldn't it make more sense to have a separate handler interface for iterable side inputs?

The idea is that for side inputs, you can create something which delegates based upon the access pattern so you effectively have three `StateRequestHandlers`, one for iterable, one for multimap, and one that delegates based upon access pattern.

Reply:
I actually started down this road, but found myself duplicating a huge amount of code due to the lack of a common interface (short of passing the full StateRequest which seemed to break the abstraction). That's not to say it wouldn't be worth exploring more, but it seems every access pattern (geo-spacial, range-based, ...) falls naturally into the pattern of having a key that is interpreted/ignored according to the access pattern's semantics. The other alternative is to let the StateHandlerFactory interface have a handle-returning-method per-known access pattern rather than pass this down to the runners themselves, which also felt odd.

Reply:
If runners want to write one object that supports both types, runners should be able to have a single class which has shared state and exposes a trivial wrapper object that implements the desired interface that converts calls to the shared object.